### PR TITLE
Remove move RenderState/TextureState Patches

### PR DESCRIPTION
**Pull request by LukeUsher**
With this work, all of these patches are no longer required, and
instead, are replaced with a RenderStateConverter and
TextureStateConverter object

One single render state patch (D3DDevice_SetRenderState_Simple) is still
required, as that bypasses the D3D__RenderStates array in some XDK
versions.

The bulk of this is removing patches, and replacing the patching mechanism and ugly/awkward to maintain deferred state handling with nice RenderState/TextureState converter objects.

The result is the removal of 40 HLE patches.

NOTE: Draft status will be removed after regression testing, as this is a huge change.

### CODE REVIEW

**Comment by PatrickvL** [comment]
Please put the removed patches in the file we keep around for that

File: src/core/hle/Patches.cpp [hunk 1]
```diff
@@ -455,9 +415,9 @@ inline void EmuInstallPatch(std::string FunctionName, xbaddr FunctionAddr)
 		printf("HLE: %s: Skipped (LLE OHCI Enabled)\n", FunctionName.c_str());
 		return;
 	}
-
-    // HACK: Some titles require unpatched Fibers, otherwise they enter an infinte loop
-    // while others require patched Fibers, otherwise they outright crash
+
+    // HACK: Some titles require unpatched Fibers, otherwise they enter an infinte loop
```
Comment (Margen67) [inline src/core/hle/Patches.cpp]: *infinite

**Comment by gandalfthewhite19890404** [comment]
WWE RAW 2:
image

**Comment by Margen67** [comment]
Instructions and characters now (sort of) render in Sneakers.
Master:
Master
PR:
PR

**Comment by medievil1** [comment]
buffy the vampire slayer doesn't render at all (black screen) with video and game 3d audio in the background

**Comment by PatrickvL** [comment]
Please also list the xdk version of apparent regressions, so we're able to compare against titles using the same xdk version

**Comment by LukeUsher** [comment]
@gandalfthewhite19890404 please don't post screenshots without context. If the game improved, that's good, but it's not a regression.

If the game got worse, please explain how/what changed, with a comparison screenshot if possible.

Posting a screenshot with no context is unhelpful and it will probably be ignored.

**Comment by medievil1** [comment]
previous comment updated with xdk version

**Comment by gandalfthewhite19890404** [comment]
FIFA World Cup 2006 - XDK 5849:
image

File: src/core/hle/D3D8/Direct3D9/Direct3D9.cpp [hunk 2]
```diff
@@ -6791,7 +6791,15 @@ void CxbxUpdateNativeD3DResources()
 VOID __declspec(noinline) D3DDevice_SetPixelShaderCommon(DWORD Handle)
 {
     // Cache the active shader handle
-	g_D3DActivePixelShader = (XTL::X_PixelShader*)Handle;
+    g_D3DActivePixelShader = (XTL::X_PixelShader*)Handle;
+
+    // Copy the Pixel Shader data to our RenderState handler
+    // This mirrors the fact that unpathed SetPixelShader does the same thing!
```
Comment (RadWolfie) [inline src/core/hle/D3D8/Direct3D9/Direct3D9.cpp]: unpatched* 😉

**Comment by jackchentwkh** [comment]
In Otogi, when you pressed A in menu, there will be a water wave effect. The effect is broken now. I guess it's a regression.

**Comment by LukeUsher** [comment]
@jackchentwkh if it happens in the development build also, then it doesn't count as part of this PR
But it should be looked into as part of another task.

**Comment by PatrickvL** [comment]
@gandalfthewhite19890404 Could you re-test with the latest build? (PS Your screenshots are somehow not showing, but their links do work.)

**Comment by LukeUsher** [comment]
I went ahead and updated @gandalfthewhite19890404's post to fix the broken screenshot links.

**Comment by gandalfthewhite19890404** [comment]
Tested with de2b891
- **FIFA World Cup 2006** <details> <summary>Fixed rainbow graphics </summary> 
image
</details>

- **Lord of the rings Third age** - now crashing just after start:
Received Exception Code 0xC0000005 @ EIP := 0x040CCFA9

- **The Elder Scrolls Morrowind** <details> <summary>Fixed rainbow water </summary>
image
</details>

File: src/core/hle/D3D8/Direct3D9/Direct3D9.cpp [hunk 3]
```diff
@@ -6819,7 +6820,7 @@ VOID WINAPI XTL::EMUPATCH(D3DDevice_SetPixelShader_0)
 	// Call the Xbox function to make sure D3D structures get set
 	XB_trampoline(VOID, WINAPI, D3DDevice_SetPixelShader_0, ());
     __asm {
-        push eax
+        mov eax, Handle
```
Comment (LukeUsher) [inline src/core/hle/D3D8/Direct3D9/Direct3D9.cpp]: This function (as documented in the patch comments above) accepts it's parameter in eax
For some reason, we were putting eax on the stack and passing it that way?

That was obviously never going to work:
1. The above call to XB_trampoline would have trashed eax
2. The function doesn't check the stack, the value contained within Handle should be placed in EAX before the call.

**Comment by LukeUsher** [comment]
Draft status removed: I think this is pretty much ready.

Notable improvments:
Super Monkeyball now works in-game
Vex has improved
The Incredible Hulk has improved

**Comment by Voxel9** [comment]
DolphinClassic is missing distance fog.

Master:
image

This PR:
image

**Comment by LukeUsher** [comment]
@Voxel9 which version of DolphinClassic? Many XDKs have a DolphinClassic sample, so the XDK version you're using is important.

**Comment by LukeUsher** [comment]
So it would appear that some LTCG games are experiencing new crash issues with this PR, however the behavior is weird.

Both Star Wars Battlefront II and Lord of the Rings: The Third Age crash on builds compiled with appveyor, but don't crash on builds I compile locally.

After investigating further, it appears that the issue is compiler version: Builds compiled with recent versions of Visual Studio 2019 will crash in this games, but earlier versions of 2019 and all versions of 2017 will not crash.

Compiler optimisation is probably fucking us over, needs further investigation.

**Comment by LukeUsher** [comment]
Added a fix for the crash caused within the LTCG pixel shader patch.
A similar fix should be applied to other patches later (as a seperate PR), but this is enough to avoid the crash in LOTR: The Third Age.

A seperate PR should address the crashing because it's unrelated to this PR: The actual cause is a change in AppVeyor's compiler version.

**Comment by gandalfthewhite19890404** [comment]
TES 3 Morrowind now shows dirty disc error after you start new game.

**Comment by PatrickvL** [comment]
This PR addresses part of #193 (linking for future reference)

**Pull request merged**